src/jit/jithashtable.h

```diff
@@ -0,0 +1,672 @@
+    }
+
+    // See if it is OK to grow the hash table by one element.  If not, reallocate
+    // the hash table.
+    void CheckGrowth()
+    {
+        if (m_tableCount == m_tableMax)
+        {
+            Grow();
+        }
+    }
+
+public:
+    // Reallocates a hash table to a specific size.  The size must be big enough
+    // to hold all elements in the table appropriately.
+    //
+    // Note that the actual table size must always be a prime number; the number
+    // passed in will be upward adjusted if necessary.
+    void Reallocate(unsigned newTableSize)
+    {
+        assert(newTableSize >=
+               (GetCount() * Behavior::s_density_factor_denominator / Behavior::s_density_factor_numerator));
+
+        // Allocation size must be a prime number.  This is necessary so that hashes uniformly
+        // distribute to all indices, and so that chaining will visit all indices in the hash table.
+        JitPrimeInfo newPrime = NextPrime(newTableSize);
+        newTableSize          = newPrime.prime;
+
+        Node** newTable = (Node**)m_alloc->ArrayAlloc(newTableSize, sizeof(Node*));
+
+        for (unsigned i = 0; i < newTableSize; i++)
+        {
+            newTable[i] = nullptr;
+        }
+
+        // Move all entries over to new table (re-using the Node structures.)
+
+        for (unsigned i = 0; i < m_tableSizeInfo.prime; i++)
+        {
+            Node* pN = m_table[i];
+            while (pN != nullptr)
+            {
+                Node* pNext = pN->m_next;
+
+                unsigned newIndex  = newPrime.magicNumberRem(KeyFuncs::GetHashCode(pN->m_key));
+                pN->m_next         = newTable[newIndex];
+                newTable[newIndex] = pN;
+
+                pN = pNext;
+            }
+        }
+
+        // @todo:
+        // We might want to try to delay this cleanup to allow asynchronous readers
+        if (m_table != nullptr)
+        {
+            m_alloc->Free(m_table);
+        }
+
+        m_table         = newTable;
+        m_tableSizeInfo = newPrime;
+        m_tableMax =
+            (unsigned)(newTableSize * Behavior::s_density_factor_numerator / Behavior::s_density_factor_denominator);
+    }
+
+    // For iteration, we use a pattern similar to the STL "forward
+    // iterator" pattern.  It basically consists of wrapping an
+    // "iteration variable" in an object, and providing pointer-like
+    // operators on the iterator. Example usage:
+    //
+    // for (JitHashTable::KeyIterator iter = foo->Begin(), end = foo->End(); !iter.Equal(end); iter++)
+    // {
+    //      // use foo, iter.
+    // }
+    // iter.Get() will yield (a reference to) the
+    // current key.  It will assert the equivalent of "iter != end."
+    class KeyIterator
+    {
+    private:
+        friend class JitHashTable;
+
+        // The method implementations have to be here for portability.
+        // Some compilers won't compile the separate implementation in shash.inl
```

Comment:
You should probably update this comment, perhaps "Some compilers won't compile a separate implementation"?